l1-contracts/test/merkle/UnbalancedMerkle.t.sol

```diff
@@ -0,0 +1,221 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright 2024 Aztec Labs.
+pragma solidity >=0.8.18;
+
+import {Test} from "forge-std/Test.sol";
+import {Hash} from "../../src/core/libraries/Hash.sol";
+
+import {TxsDecoderHelper} from "../decoders/helpers/TxsDecoderHelper.sol";
+/**
+ * Tests the tree construction for unbalanced rollups.
+ * Used for calculating txsEffectsHash over non balanced rollups - each leaf is one baseLeaf
+ * calculated in TxsDecoder.sol.
+ */
+
+contract UnbalancedMerkleTest is Test {
+  /**
+   * Rollups are constructed greedily, with a set of N transactions split into subtrees of decreasing
+   * powers of 2.
+   * We list them in reverse order as we compute subtree roots from R to L
+   */
+  TxsDecoderHelper internal txsHelper;
+
+  function setUp() public {
+    txsHelper = new TxsDecoderHelper();
+  }
+
+  function testDecomp() public {
+    // Worst case - max num txs
+    uint32 numTxs = 65535;
+    bytes memory subtrees = txsHelper.computeSubtreeSizes(numTxs);
+    assertEq(subtrees, hex"8000400020001000080004000200010000800040002000100008000400020001");
+    (uint256 min, uint256 max) = txsHelper.computeMinMaxPathLength(numTxs);
+    assertEq(min, 15);
```

Comment:
Why is the min not 1?

Reply:
The RHS subtree is itself an unbalanced tree - not a node of 1, as we greedy fill left to right.
Like for 7, filling left to right means even though we do have a subtree of 1, the min path length is 2. As we connect it to the subtree of 2, which is then connected to the subtree of 4:
```
  //                     root
  //             /                  \
  //         merge3                merge5
  //      /          \             /    \
  //   merge1       merge2      merge4  base
  //  /     \      /     \      /    \
  // base  base  base   base  base  base
```